icu4c/source/common/uloc.cpp

```diff
@@ -1588,77 +1553,85 @@ _canonicalize(const char* localeID,
 
     origLocaleID=tmpLocaleID;
 
-    /* if we are doing a full canonicalization, then put results in
-       localeBuffer, if necessary; otherwise send them to result. */
-    if (/*OPTION_SET(options, _ULOC_CANONICALIZE) &&*/
-        (result == NULL || resultCapacity < (int32_t)sizeof(localeBuffer))) {
-        name = localeBuffer;
-        nameCapacity = (int32_t)sizeof(localeBuffer);
-    } else {
-        name = result;
-        nameCapacity = resultCapacity;
+    LocalMemory<char> localeBuffer;
+    LocalPointer<CheckedArrayByteSink> localSink;
+    internal::MemoryByteSink* sinkp = &sink;
+    {
+        char sentinel;
+        int32_t dummy;
+
+        char* buffer = sink.GetAppendBuffer(
+                ULOC_FULLNAME_CAPACITY,
+                ULOC_FULLNAME_CAPACITY,
+                &sentinel,
+                ULOC_FULLNAME_CAPACITY,
+                &dummy);
+
+        /* if we are doing a full canonicalization, then put results in
+           localeBuffer, if necessary; otherwise send them directly to sink. */
+        if (/*OPTION_SET(options, _ULOC_CANONICALIZE) &&*/
+            buffer == &sentinel) {
+            if (localeBuffer.allocateInsteadAndReset(
+                    ULOC_FULLNAME_CAPACITY) == nullptr) {
```

Comment:
For a small, fixed-sized array we can just use stack-allocation.

On the other hand, why not CharString + CharStringByteSink?